source/adapters/cuda/context.hpp

```diff
@@ -88,14 +117,29 @@ struct ur_context_handle_t_ {
 
   ur_context_handle_t_(const ur_device_handle_t *Devs, uint32_t NumDevices)
       : Devices{Devs, Devs + NumDevices}, RefCount{1} {
+    CUevent EvBase = nullptr;
+    int i = 0;
     for (auto &Dev : Devices) {
       urDeviceRetain(Dev);
+      Dev->retainNativeContext();
```

Comment:
I would personally prefer to see all of this logic for contexts happen in `ur_queue_handle_t_`s. This avoids giving `sycl::context`s extra semantics for the CUDA backend. Within `urQueueCreate` you could call something like `ur_device_handle_t_::init_device()` which would retain the primary ctx and then set the base event, which would then be cached in the device, so if another queue is created for the same device, it doesn't need to do the same base event getting, info querying, etc.

Reply:
Doing it in the queue would be nicer, it means we wouldn't need the user code changes dealing with `sycl::context` since the queue already works with just one device. Not 100% sure if would work well with the event timing though.

Reply:
OK I'll close this PR. I've described how it would work with event timings and any other edge case limitations when using the `queue`/`kernel_bundle`/etc to instantiate `CUcontext` here: https://github.com/oneapi-src/unified-runtime/pull/2077#issuecomment-2346418393
An third alternative of creating it on demand and allowing the number of `CUcontext` to grow to \infty and then relying on final dpc++ runtime to reset `CUcontext` has similar edge case limitations in addition to the question of whether `CUcontext` references can grow to \infty without consequences.